### ISSUE TITLE
Can't add header to request using the method: withHeader(), withAddedHeader()

### ISSUE THREAD

**Issue by denormalizer** [guzzle/psr7#176]
**TLDR;**
I can't add header to request using the method: `withHeader()`, `withAddedHeader()`

**Long**

Guzzle 5 documentation suggests it can be added like this:

http://docs.guzzlephp.org/en/5.3/quickstart.html

    // Set a single value for a header
    $request->setHeader('User-Agent', 'Testing!');
    
    // Set multiple values for a header in one call
    $request->setHeader('X-Foo', ['Baz', 'Bar']);
    
    // Add a header to the message
    $request->addHeader('X-Foo', 'Bam');

However, in Guzzle 6, this documentation seems to be missing.

The only way documented is to pass through the constructor:

http://docs.guzzlephp.org/en/stable/request-options.html#headers

    use GuzzleHttp\Psr7\Request;
    $request = new Request('GET', 'http://foo.com', ['X-Foo' => 'test']);

Looking at the class methods available, I see two new methods which may be doing the same thing:

    [13] => withHeader
    [14] => withAddedHeader

These new methods didn't produce the results I was expecting:

    use GuzzleHttp\Psr7\Request;
    $request =  new Request('GET', $uri);
    $request->withHeader('Encoding', 'gzip,deflate');
    $request->withAddedHeader('Content-Type', 'application/json');
    print_r($request->getHeaders());
    // PRINTS
    Array
    (
        [Host] => Array
            (
                [0] => api.host.com
            )
    
    )

Should I be using them in a different way?

Any help would be appreciated. Thanks.

<HR>

Guzzle: Guzzle v6.3

**Comment by jeskew** [comment]
Hi @denormalizer,

Guzzle 6 follows [the PSR-7 specification](http://www.php-fig.org/psr/psr-7/), which mandates that all HTTP messages be immutable. When you call `withHeader` or `withAddedHeader`, the method's return value will be a message with the requested header changes applied. The message on which the `with*` method was called will not be modified.

**Comment by denormalizer** [comment]
Thanks for the clarification @jeskew
